### PR TITLE
LACP support

### PR DESCRIPTION
**Pull request by ydirson**
NOTE: [submitted upstream](https://github.com/xenserver/host-installer/pull/37)

This branch deals with the installer side of things, firstboot data needs to be handled in XAPI, for which we depend on https://github.com/xapi-project/xen-api/pull/4333:
- configuration from cmdline (covers fetching answerfile and the next items)
- configuration from answerfile
- configuration UI for netinstall
- configuration UI of mgmt interface, writing data for firstboot

Known limitations we may want to address:
- bad cmdline config could be reported cleanly (we currently get a python backtrace, which gets garbled by the TUI)
- does not deal with DebStyleInterface, ie. does not support Debian-based installer; shouldn't that code be simply be nuked ?
- limitations that could be addressed by moving bond creation/management to a preliminar "advanced networking" step similar to "advanced storage" (adding a new F key for this to initial screen may look bad, though):
  - bond interface is not created immediately, so appears in the interface list as "[no link]". Maybe this can be dealt with, by first creating `ifcfg-*` files without an IP config.
  - cannot change our mind once the eth* to create the bond have been made

This PR explicitly does not deal with:
- creating more than 1 bond
- system upgrade parsing during upgrade (get mgmt-iface config in `ExistingInstallation.readSettings()`), to which `network_db` seems to be unable to report about bonding (but the impact seems to be limited to default values).  The system's bonds are already preserved on upgrade, though.

What the screens currently look like:

lacp-0-ifaces
lacp-1-create
lacp-2-confirm
lacp-3-ifaces

### CODE REVIEW

File: netutil.py [hunk 1]
```diff
@@ -20,10 +20,20 @@ def __init__(self, nic_dict):
...
 
+    @classmethod
+    def create_bonding(cls, devname, mac, bond_mode, bond_members):
+        nic = NIC({"Kernel name": devname,
+                   "Assigned MAC": mac,
+                   })
+        nic.bond_mode = bond_mode
+        nic.bond_members = bond_members
+        return nic
+
```
Comment (benjamreis) [inline netutil.py]: You don't use `cls` why not make a `@staticmethod` then?

**Comment by benjamreis** [comment]
https://github.com/xapi-project/xen-api/pull/4333

File: netutil.py [hunk 2]
```diff
@@ -45,16 +45,46 @@ def scanConfiguration():
...
+    for iface in virtual_devs:
+        mode_file = os.path.join(SYSVIRTUALNET, iface, "bonding/mode")
+        if not os.path.exists(mode_file):
+            logger.log("scanConfiguration: no file %r" % mode_file)
+            continue            # not a bonding iface
+        with open(mode_file) as fd:
+            lines = fd.readlines()
+            assert len(lines) == 1
+            if lines[0].strip() != "802.3ad 4":
+                logger.log("scanConfiguration: wrong bonding mode %r" % lines[0])
```
Comment (stormi) [inline netutil.py]: Would it be useful to specify in the log that the expected bonding mode is LACP?

File: tui/network.py [hunk 3]
```diff
@@ -159,6 +159,44 @@ def iface_details(context, conf):
     tui.screen.popHelpLine()
     return True
 
+def lacp_bond_ui(conf):
```
Comment (stormi) [inline tui/network.py:162]: Next time you test it, could you take screenshots of the UI?

File: netutil.py (hunk 1 above)
Comment (ydirson) [reply netutil.py]: right, but this is going away anyawy

File: tui/network.py (hunk 3 above)
Comment (ydirson) [reply tui/network.py:162]: ok, when it's ready enough ;)

**Comment by ydirson** [comment]
Working on answerfile shows exacerbated need to pass a NIC object to NetInterface ctor (instead of "just hwaddr" initially, getting to which we now add bonding parameters).  The only call sites for which conversion is still not obvious to me are the ones in `product.py`, dealing with the upgrade process.

**Pull request closed without merging**